README.md
# Haskell Binary String Evolution

See writeup [here](https://medium.com/knoxcs/binary-string-evolution-in-haskell-574afbbcdc59).

## Problem

Evolve the largest number you can represent with a 20-bit string.

## Implementation

### 20 Bit Strings

Instead of storing an array of 20 1s and 0s, the `Word32` type from `Data.Word` 
can be used to store the relevant data. All calculations with the `Word32` type
will be made such that the other 12 bits will always be 0. This allows for the
bitcount of the `Word32` to be the fitness function itself and allows for the use
of bitwise operations to manipulate the 1s and 0s. 

### Crossover

Crossover is implemented using the standard crossover technique, where both
parents are split into two parts and merged to create a child. Thanks to
bitwise operations, this process is not too difficult. Below is the algorithm
used to calculate the child's DNA. Note that this algorithm assumes the largest
12 bits of both parents are 0, which allows for no need to modify the 12 bits
after the crossover. Crossover happens always.

```
splitPoint = random int between 0 and 20 inclusive
mask = ~0 << splitPoint
child = (parent1 & mask) | (parent2 & ~mask)
```

### Mutation

For each bit in the 20 bits of the `Word32`, there is a small chance (0.1%)
that it will be flipped after crossover occurs. The bit is flipped with bitwise
xor. Mutation always happens after crossover occurs.

```
n = current bit # (between 0-20 exclusive)
child ^= (1 << n)
```

### Population Selection & Reproduction

When parents are selected from the population, a roulette selection is used to
bias the selection towards candidates with a higher fitness.

To create the next generation, two parents are selected from the population,
crossed, and then mutated which yields the child for the next generation.

## Results

### Visualization

Using the matplotlib bindings for Haskell, a simple visualization can be
generated that shows how different parameters affect the efficiency of the
genetic algorithm.

Here are the details for each experiement:

|  | Mutation Rate | Crossover Rate | Crossover Type | Elitism | Population Size |
|---|---|---|---|---|---|
| EXP 0 | 0% | 100% | Single Point | Top 25 | 50 |
| EXP 1 | 100% | 80% | Single Point | Top 1 | 50 |
| EXP 2 | 1% | 20% | Single Point | None | 50 |
| EXP 3 | 50% | 80% | Single Point | Top 1 | 50 |
| EXP 4 | 10% | 100% | Single Point | 0 | 100 |

And here is a graph of the results showing how long it took each strategy to
reach the optimal solution (all 1s):

![Graph](https://raw.githubusercontent.com/Archaversine/HaskellBinEvo/main/visualization.svg) 

### Results from an Efficient Strategy

Sample results from running the program:

```
--------------------------------------------
Population Size: 500
Number of Generations: 30
Number of Iterations: 30
Mutation per bit: 0.1%
--------------------------------------------
Iteration 1: 11111111111111111111 = 1048575
Iteration 2: 11111111111111111111 = 1048575
Iteration 3: 11111111111111111111 = 1048575
Iteration 4: 11111111111111111111 = 1048575
Iteration 5: 11111111111111111111 = 1048575
Iteration 6: 11111111111111111111 = 1048575
Iteration 7: 11111111111111111111 = 1048575
Iteration 8: 11111111111111111111 = 1048575
Iteration 9: 11111111111111111111 = 1048575
Iteration 10: 11111111111111111111 = 1048575
Iteration 11: 11111111111111111111 = 1048575
Iteration 12: 11111111111111111111 = 1048575
Iteration 13: 11111111111111111111 = 1048575
Iteration 14: 11111111111111111111 = 1048575
Iteration 15: 11111111111111111111 = 1048575
Iteration 16: 11111111111111111111 = 1048575
Iteration 17: 11111111111111111111 = 1048575
Iteration 18: 11111111111111111111 = 1048575
Iteration 19: 11111111111111111111 = 1048575
Iteration 20: 11111111111111111111 = 1048575
Iteration 21: 11111111111111111111 = 1048575
Iteration 22: 11111111111111111111 = 1048575
Iteration 23: 11111111111111111111 = 1048575
Iteration 24: 11111111111111111111 = 1048575
Iteration 25: 11111111111111111111 = 1048575
Iteration 26: 11111111111111111111 = 1048575
Iteration 27: 11111111111111111111 = 1048575
Iteration 28: 11111111111111111111 = 1048575
Iteration 29: 11111111111111111111 = 1048575
Iteration 30: 11111111111111111111 = 1048575
--------------------------------------------
Minimum Fitness: 20.0
Maximum Fitness: 20.0
Average Fitness: 20.0
Standard Deviation: 0.0
--------------------------------------------
```



CHANGELOG.md
# Changelog for `BinaryEvolution`

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to the
[Haskell Package Versioning Policy](https://pvp.haskell.org/).

## Unreleased

## 0.1.0.0 - YYYY-MM-DD


src/Main.hs
module Main (main) where

import Data.Word (Word32)
import Data.Bits (shiftL, shiftR, xor, complement, popCount, (.|.), (.&.))
import Data.List (sortBy)
import Data.Functor ((<&>))

import Control.Monad (replicateM)

import Graphics.Matplotlib

import System.Random (randomRIO)

type Genome = Word32
type Population = [Genome]
type ScoredPopulation = [(Genome, Double)]

type Point = (Int, Int)

data CrossOverType = SinglePoint | Uniform deriving (Eq, Show)

data GASettings = GASettings {
    mutationRate :: Double,
    crossoverRate :: Double,
    crossoverType :: CrossOverType,
    elitism :: Int,
    popSize :: Int,
    gens :: Int
} deriving Show

-- Variables

-- Number where only 20/32 bits are 1
-- Used to set max limit for random generation
mask :: Genome
mask = 1048575

-- Debugging Functions
-- Only prints 20 bits (because we are only interacting with 20)
printBits :: Genome -> IO ()
printBits w = do
    mapM_ (\n -> putStr . show $ 1 .&. (w `shiftR` n)) [19,18..0] 
    putStrLn $ " = " ++ show w

-- Genetic Algorithm Functions

genRandomGenome :: IO Genome
genRandomGenome = randomRIO (0, mask)

genInitialPopulation :: GASettings -> IO Population
genInitialPopulation settings = replicateM (popSize settings) genRandomGenome

fitness :: Genome -> Double
fitness = fromIntegral . popCount

cross :: GASettings -> Genome -> Genome -> IO Genome
cross settings p1 p2
    | crossoverType settings == SinglePoint = singlePointCross p1 p2
    | otherwise = uniformCross p1 p2

singlePointCross :: Genome -> Genome -> IO Genome
singlePointCross p1 p2 = do
    crossPoint <- randomRIO (0, 20) :: IO Int
    let crossMask = complement 0 `shiftL` crossPoint

    return $ (p1 .&. crossMask) .|. (p2 .&. complement crossMask)

uniformCross :: Genome -> Genome -> IO Genome
uniformCross p1 p2 = do
    m <- randomRIO (0, mask) :: IO Genome
    return $ (p1 .&. m) .|. (p2 .&. complement m)

mutate :: GASettings -> Genome -> IO Genome
mutate settings g = do
    r <- randomRIO (0, 1) :: IO Double
    point <- randomRIO (0, 10) :: IO Int

    if r > mutationRate settings then return g
        else return $ g `xor` (1 `shiftL` point)

-- Cannot handle empty list as input
selectFromPopulation :: ScoredPopulation -> Double -> IO Genome
selectFromPopulation scored totalFitness = do
    r <- randomRIO (0, 1) :: IO Double
    return $ go (r * totalFitness) scored
        where go _ [] = undefined
              go _ [(x, _)] = x
              go dart ((x, p):xs)
                | dart > p = go dart xs
                | otherwise = x

reproduce :: GASettings -> Population -> IO Population
reproduce settings p = do
    let elite = take (elitism settings) p
    nonElite <- reproduceNonElite settings p
    return $ elite ++ nonElite

reproduceNonElite :: GASettings -> Population -> IO Population
reproduceNonElite settings p = replicateM (popSize settings - elitism settings) $ do
    p1 <- selectFromPopulation scored totalFitness
    p2 <- selectFromPopulation scored totalFitness

    crossChance <- randomRIO (0, 1) :: IO Double

    if crossChance > crossoverRate settings then return p1
        else cross settings p1 p2 >>= mutate settings
    --cross settings p1 p2 >>= mutate settings
        where summed = scanl1 (+) $ map fitness p :: [Double]
              scored = zip p summed
              totalFitness = last summed

-- Returns best genome
runGA :: GASettings -> IO Genome
runGA settings = go (gens settings) (genInitialPopulation settings)
    where nextGen = reproduce settings
          go 0 pop = pop <&> head -- Return first genome of population (list should be sorted by now)
          go n pop = go (n - 1) $ pop >>= nextGen <&> sortBy (flip compare)

runUntilOptimal :: GASettings -> IO Int
runUntilOptimal settings = go (genInitialPopulation settings) 0
    where nextGen = reproduce settings
          go pop n = do
            candidates <- pop
            if head candidates /= mask then go (pop >>= nextGen <&> sortBy (flip compare)) (n + 1)
                else return n
    

visualize :: [[Point]] -> Matplotlib
visualize [] = undefined
visualize points = go points 0 $ title "Genetic Algorithm Experiments"
    where go :: [[Point]] -> Int -> Matplotlib -> Matplotlib
          go [] _ acc = acc
          go (pt:pts) n acc = go pts (n + 1) $ acc % line xs ys @@ [o2 "label" ("EXP " ++ show n)]
            where (xs, ys) = unzip pt

main :: IO ()
main = do
    let runs = 30

        test1 = GASettings { mutationRate = 0, 
                             crossoverRate = 1, 
                             crossoverType = SinglePoint, 
                             elitism = 25, 
                             popSize = 50,
                             gens = 0 }

        test2 = GASettings { mutationRate = 1, 
                             crossoverRate = 0.8, 
                             crossoverType = SinglePoint, 
                             elitism = 1, 
                             popSize = 50,
                             gens = 0 }

        test3 = GASettings { mutationRate = 0.01,
                             crossoverRate = 0.2,
                             crossoverType = SinglePoint,
                             elitism = 0,
                             popSize = 50,
                             gens = 0 }

        test4 = GASettings { mutationRate = 0.5,
                             crossoverRate = 0.8,
                             crossoverType = SinglePoint,
                             elitism = 1,
                             popSize = 50,
                             gens = 0 }

        test5 = GASettings { mutationRate = 0.1,
                             crossoverRate = 1,
                             crossoverType = Uniform,
                             elitism = 0,
                             popSize = 100,
                             gens = 0 }

    results1 <- replicateM runs $ runUntilOptimal test1
    putStrLn "Finished EXP 0"

    results2 <- replicateM runs $ runUntilOptimal test2
    putStrLn "Finished EXP 1"

    results3 <- replicateM runs $ runUntilOptimal test3
    putStrLn "Finished EXP 2"

    results4 <- replicateM runs $ runUntilOptimal test4 
    putStrLn "Finished EXP 3"

    results5 <- replicateM runs $ runUntilOptimal test5 
    putStrLn "Finished EXP 4"

    let points1 = zip [1..] results1
        points2 = zip [1..] results2
        points3 = zip [1..] results3
        points4 = zip [1..] results4
        points5 = zip [1..] results5

        img = visualize [points1, points2, points3, points4, points5]

    svg <- toSvg $ img % xlabel "Run #" 
                       % ylabel "Generations to Optimal"
                       % legend

    case svg of
        Left err -> putStrLn $ "Error: " ++ err
        Right s -> writeFile "visualization.svg" s

--main :: IO ()
--main = do 
--    putStrLn "--------------------------------------------"
--    putStrLn $ "Population Size: " ++ show population
--    putStrLn $ "Number of Generations: " ++ show generations
--    putStrLn $ "Number of Iterations: " ++ show iterations
--    putStrLn $ "Mutation per bit: " ++ show (mutationChance * 100) ++ "%"
--    putStrLn "--------------------------------------------"
--
--    results <- go iterations (pure [])
--
--    let floats = map fromIntegral results :: [Double]
--        mean = sum floats / fromIntegral iterations
--        totalDiffSquared = sum $ map (\x -> (x - mean) * (x - mean)) floats
--        deviation = sqrt $ totalDiffSquared / fromIntegral iterations
--
--    putStrLn "--------------------------------------------"
--    putStrLn $ "Minimum: " ++ show (minimum results)
--    putStrLn $ "Maximum: " ++ show (maximum results)
--    putStrLn $ "Average: " ++ show mean
--    putStrLn $ "Standard Deviation: " ++ show deviation
--    putStrLn "--------------------------------------------"
--
--    where go 0 acc = acc
--          go n acc = do
--            best <- runGA
--            putStr ("Iteration " ++ show i ++ ": ") >> printBits best
--            go (n - 1) $ (best :) <$> acc
--                where i = iterations - n + 1
